python_modules/libraries/dagster-dbt/dagster_dbt/core/resources_v2.py

```diff
@@ -241,40 +260,58 @@ def to_default_asset_events(
-                severity = (
-                    AssetCheckSeverity.WARN
-                    if node_status == TestStatus.Warn
-                    else AssetCheckSeverity.ERROR
-                )
-
                 attached_node_resource_props: Dict[str, Any] = manifest["nodes"].get(
                     attached_node_unique_id
                 ) or manifest["sources"].get(attached_node_unique_id)
                 attached_node_asset_key = dagster_dbt_translator.get_asset_key(
                     attached_node_resource_props
                 )
+                check_name = event_node_info["node_name"]
+
+                if (
+                    context
+                    and AssetCheckKey(asset_key=attached_node_asset_key, name=check_name)
+                    in context.selected_asset_check_keys
+                ):
+                    is_test_successful = node_status == TestStatus.Pass
+                    severity = (
+                        AssetCheckSeverity.WARN
+                        if node_status == TestStatus.Warn
+                        else AssetCheckSeverity.ERROR
+                    )
 
-                yield AssetCheckResult(
-                    passed=is_test_successful,
-                    asset_key=attached_node_asset_key,
-                    check_name=event_node_info["node_name"],
-                    metadata=metadata,
-                    severity=severity,
-                )
-            else:
-                for upstream_unique_id in upstream_unique_ids:
-                    upstream_resource_props: Dict[str, Any] = manifest["nodes"].get(
-                        upstream_unique_id
-                    ) or manifest["sources"].get(upstream_unique_id)
-                    upstream_asset_key = dagster_dbt_translator.get_asset_key(
-                        upstream_resource_props
+                    yield AssetCheckResult(
+                        passed=is_test_successful,
+                        asset_key=attached_node_asset_key,
+                        check_name=check_name,
+                        metadata=metadata,
+                        severity=severity,
                     )
 
-                    yield AssetObservation(
-                        asset_key=upstream_asset_key,
+                # dbt's default indirect selection (eager) will select relationship tests
+                # on unselected assets, if they're compared with a selected asset.
+                # This doesn't match Dagster's default check selection which is to only
+                # select checks on selected assets. When we use eager, we may receive
+                # unexpected test results so we log those as observations as if
+                # asset checks were disabled.
+                else:
+                    logger.info(
+                        f"dbt test '{check_name}' will be logged as an AssetObservation instead of "
+                        "an AssetCheckResult because the asset check was not selected."
```

Comment:
Logging a message per conversion from `AssetCheckResult` to `AssetObservation` could be a bit spammy, but I agree it's good to have this in the event log.

Though if it shows up in the description of `AssetObservation`, then lets go with that.

Reply:
It shows up on the asset page but not in the event log, so I'll keep this log

![Screenshot 2024-03-24 at 7 50 11 PM](https://github.com/dagster-io/dagster/assets/22018973/6aef5059-9564-4300-a6a3-73dcdc6442fc)
![Screenshot 2024-03-24 at 7 50 15 PM](https://github.com/dagster-io/dagster/assets/22018973/0d5a1e88-6be6-4243-b85c-7faea2a9ba34)